tapdb/universe_perf_test.go

```diff
@@ -0,0 +1,679 @@
+		t.Logf("  Without indices: %v", result.withoutIndices)
+		t.Logf("  Improvement:     %.2fx", improvement)
+	}
+}
+
+// Helper for sorting checks
+func isSortedWithDirection(s []universe.AssetSyncSnapshot, sortType universe.SyncStatsSort, 
+	direction universe.SortDirection) func(i, j int) bool {
+	
+	asc := direction == universe.SortAscending
+
+	return func(i, j int) bool {
+		switch sortType {
+		case universe.SortByAssetName:
+			if asc {
+				return s[i].AssetName < s[j].AssetName
+			}
+			return s[i].AssetName > s[j].AssetName
+
+		case universe.SortByTotalSyncs:
+			if asc {
+				return s[i].TotalSyncs < s[j].TotalSyncs
+			}
+			return s[i].TotalSyncs > s[j].TotalSyncs
+
+		case universe.SortByTotalProofs:
+			if asc {
+				return s[i].TotalProofs < s[j].TotalProofs
+			}
+			return s[i].TotalProofs > s[j].TotalProofs
+
+		default:
+			return false
+		}
+	}
+}
+
+func TestUniversePerfInserts(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping insert performance test in short mode")
+	}
+
+	t.Parallel()
+
+	const (
+		numAssets = 50
+		numLeavesPerTree = 10
+		numEventsPerAsset = 25
+	)
+
+	var batchSizes = []int{1, 5, 10}
+
+	type insertStats struct {
+		name           string
+		withoutIndices time.Duration
+		withIndices    time.Duration
+		batchSize      int
+		totalInserts   int
+	}
+	testResults := make(map[string]*insertStats)
+
+	// Function to run performance tests with and without indices
+	runTest := func(withIndices bool) {
+		t.Run(fmt.Sprintf("indices=%v", withIndices), func(t *testing.T) {
+			ctx := context.Background()
+			db := NewTestDB(t)
+			
+			// Drop or create indices based on test scenario
+			sqlDB := db.BaseDB
+			if !withIndices {
+				t.Log("Dropping indices...")
+				_, err := sqlDB.Exec(`
+					DROP INDEX IF EXISTS idx_universe_roots_asset_group_proof;
+					DROP INDEX IF EXISTS idx_universe_events_type_counts;
+					DROP INDEX IF EXISTS idx_universe_events_universe_root_id;
+					DROP INDEX IF EXISTS idx_universe_events_sync;
+					DROP INDEX IF EXISTS idx_universe_leaves_universe_root_id;
+					DROP INDEX IF EXISTS idx_universe_leaves_asset_genesis_id;
+				`)
+				require.NoError(t, err)
+			} else {
+				t.Log("Creating indices...")
+				_, err := sqlDB.Exec(`
+					CREATE INDEX IF NOT EXISTS idx_universe_roots_asset_group_proof 
+					ON universe_roots (asset_id, group_key, proof_type);
+					CREATE INDEX IF NOT EXISTS idx_universe_events_type_counts 
+					ON universe_events (event_type, universe_root_id);
+					CREATE INDEX IF NOT EXISTS idx_universe_events_universe_root_id 
+					ON universe_events (universe_root_id);
+					CREATE INDEX IF NOT EXISTS idx_universe_events_sync 
+					ON universe_events (event_type);
+					CREATE INDEX IF NOT EXISTS idx_universe_leaves_universe_root_id 
+					ON universe_leaves (universe_root_id);
+					CREATE INDEX IF NOT EXISTS idx_universe_leaves_asset_genesis_id 
+					ON universe_leaves (asset_genesis_id);
+				`)
+				require.NoError(t, err)
+			}
```

Comment:
This should probably be a helper function since it looks like the same code is used three times (but different versions of it?)